specs/packagemanager/PackageManagement.md

```diff
@@ -0,0 +1,1136 @@
+* EnsureIsReady -- Is a package ready for use and, if not, make it so
+* IsPackageRegistrationPending -- Is there an update waiting to install?
+* PackageSets -- Batch operations
+* PackageRuntimeManager -- Batch operations for use at runtime via Dynamic Dependencies
+* Usability -- Quality-of-Life enhancements
+
+## 3.1. API Structure
+
+Methods to drive deployment activity typically follow the pattern:
+
+    <verb>Package[Set][By<Type>][Suffix]Async(target, options)
+
+The following verbs are supported:
+
+* Is...Ready
+* Ensure...IsReady
+* Add
+* Stage
+* Register
+* Remove
+* Repair
+* Reset
+* Provision
+* Deprovision
+
+These methods accept their target package(s) as their first parameter. This parameter can be an
+various types of information including:
+
+* Filename
+* Path
+* PackageFamilyName
+* PackageFullName
+* Uri
+
+**NOTE:** Methods with a target of a URI are named ...ByUri....
+
+**NOTE:** Methods with a target of a PackageFamilyName are named ...ByPackageFamilyName....
+
+**NOTE:** Methods with a target of a PackageFamilyName are named ...ByPackageFullName....
+
+**NOTE:** Methods with no ...By<Type>... qualifier accept a string which can contain one or more types
+of identifiers. These can vary for different verbs. See the per-method documentation for the
+specific target types supported by each method.
+
+These methods accept options as a matching `<verb>Package[Set]Options` type, e.g.
+`AddPackageAsync(string packageUri, AddPackageOptions options)`.
+
+The following table shows the supported permutations of verbs and targets:
+
+|Verb         | Path | Filename | PackageFamilyName | PackageFullName | file:  | http(s): | ms-uup: | PackageSet |
+|-------------|:----:|:--------:|:-----------------:|:---------------:|:------:|:--------:|:-------:|:----------:|
+|IsReady      |  X   |    X     |       OS/WAS      |       WAS       |   X    |    X     |  WAS    |    WAS     |
+|EnsureIsReady|  X   |    X     |         X         |        X        |   X    |    X     |  WAS    |    WAS     |
+|Add          | WAS  |    X     |         X         |        X        | OS/WAS |  OS/WAS  | OS/WAS  |    WAS     |
+|Stage        | WAS  |    X     |         X         |        X        | OS/WAS |  OS/WAS  | OS/WAS  |    WAS     |
+|Register     | WAS  |  OS/WAS  |       OS/WAS      |      OS/WAS     | OS/WAS |    X     | OS/WAS  |    WAS     |
+|Remove       |  X   |    X     |        WAS        |      OS/WAS     |   X    |    X     | OS/WAS  |    WAS     |
+|Repair       |  X   |    X     |        WAS        |       WAS       |   X    |    X     |  WAS    |    WAS     |
+|Reset        |  X   |    X     |        WAS        |       WAS       |   X    |    X     |  WAS    |    WAS     |
+|Provision    |  X   |    X     |       OS/WAS      |        X        |   X    |    X     |  WAS    |    WAS     |
+|Deprovision  |  X   |    X     |       OS/WAS      |        X        |   X    |    X     |  WAS    |    WAS     |
```

Comment:
In each row, if it accepts PackageSet, is it required that all items in the PackageSet be of the same type?  Is it required that all items in the PackageSet be of types that are allowed on that row (e.g. Provision accepts only PFaN and ms-uup items in its PackageSet, but not Path or PFuN)?

Reply:
> In each row, if it accepts PackageSet, is it required that all items in the PackageSet be of the same type?

No. A PackageSet is essentially an array of package items (PackageSetItem[], in concept) and each PackageSet has its own data. There's no need for them to have any identical properties; you absolute can have packageset.PackageSetItems[0].PackageUri=file://blah/appxmanifest.xml, [1]=file://blah, [2]=ms-uup://Product/stuff.x64 and [3]=MyFamily_1234567890abc